### PR TITLE
Removal of global variable GOrbitApp

### PR DESCRIPTION
**Pull request by beckerhe**
This PR removes GOrbitApp from the code base.

It has a lot of commits - mainly separated by class. Almost all of them a straight forward.
They introduce a member variable `OrbitApp* app_` which is initialized via the constructor and used
in the class's implementation instead of GOrbitApp.

All commits compile on its own and I recommend to review them one-by-one. It's a large PR,
but the commits itself are rather small. Let me know if you think this needs to be split up into multiple PRs.

The reason for doing this is mainly #1526. This PR needs to move some resource ownership (i.e. ProcessManager) from OrbitApp to OrbitMainWindow, which means OrbitMainWindow needs to control the lifetime of OrbitApp, otherwise OrbitApp might end up with a dangling pointer when OrbitMainWindow gets out of scope. So after removing all of GOrbitApp, there is a commit that hands ownership of OrbitApp to OrbitMainWindow.

It did some extensive manual testing with the branch and I couldn't crash it or found any bug.

Note: This does not solve the cyclic dependency problem that we have with OrbitApp ("OrbitApp owns X and X needs to access OrbitApp"). It can and has to be refactored independently. This is also the reason why this change introduces a lot of forward declarations of OrbitApp. They are needed due to cyclic dependencies.

EDIT: Note that the license_headers presubmit check is expected to fail due to the known orbitcodeeditor-problem. b/163462463 for more details.

EDIT: Bug added for tracking the progress of follow-up refactorings of the cyclic dependencies: http://b/175759190

### CODE REVIEW

File: OrbitGl/App.cpp [hunk 1]
```diff
@@ -118,6 +118,11 @@ OrbitApp::OrbitApp(ApplicationOptions&& options,
 }
 
 OrbitApp::~OrbitApp() {
+  AbortCapture();
+
+  process_manager_->ShutdownAndWait();
+  thread_pool_->ShutdownAndWait();
```
Comment (dimitry-) [inline OrbitGl/App.cpp:123]: are these noexcept?

Comment (beckerhe) [reply OrbitGl/App.cpp:123]: No. Should I add a catch-all block? Maybe that's better but this would change behaviour. There was nothing that caught a potential exception before.

If these threw, this would end up in an abort since the destructor is noexcept by default.

Comment (dimitry-) [reply OrbitGl/App.cpp:123]: I did not know d-tors are noexcept by default. I am thinking it might be better to have an explicit catch all -> FATAL block,

Comment (beckerhe) [reply OrbitGl/App.cpp]: Added that. PTAL.

File: OrbitQt/orbitcodeeditor.cpp [hunk 2]
```diff
@@ -61,7 +61,6 @@
 #include <QtWidgets>
 #include <fstream>
 
-#include "App.h"
```
Comment (florian-kuebler) [inline OrbitQt/orbitcodeeditor.cpp:64]: Maybe we should remove the complete class (maybe in a different PR). WDYT?

File: OrbitGl/TracepointTrack.h [hunk 3]
```diff
@@ -7,9 +7,9 @@
 
 #include "EventTrack.h"
 
-class TracepointTrack : public EventTrack {
+class TracepointTrack final : public EventTrack {
```
Comment (florian-kuebler) [inline OrbitGl/TracepointTrack.h:10]: Any reason why this is made final? Fine with me, but seems unrelated.

**Review by florian-kuebler — APPROVED** [review]
Wow, thanks so much for this. This was on my todo list since our last fixit-week. 
On minor thing:
As you noted, we should really push forward removing the forward references here.
Can you create a bug for that and link this in this PR?

**Comment by beckerhe** [comment]
> Wow, thanks so much for this. This was on my todo list since our last fixit-week.
> On minor thing:
> As you noted, we should really push forward removing the forward references here.
> Can you create a bug for that and link this in this PR?

Hmm. I don't think I have touched a single forwarding reference. Can you explain what you're talking about? Adn why should we remove forwarding references? I'm confused. :see_no_evil:

File: OrbitGl/TracepointTrack.h (hunk 3 above)
Comment (beckerhe) [reply OrbitGl/TracepointTrack.h:10]: First I added that to help with the refactoring. Then I've seen that all the other Track classes had this final and I left it there for completeness reasons. I can remove it if you prefer.

File: OrbitQt/orbitcodeeditor.cpp (hunk 2 above)
Comment (beckerhe) [reply OrbitQt/orbitcodeeditor.cpp:64]: Yes, this will be revamped with the source code view. So there are already a variety of bugs covering that.

File: OrbitGl/TracepointTrack.h (hunk 3 above)
Comment (florian-kuebler) [reply OrbitGl/TracepointTrack.h:10]: I thought I saw other tracks that are not final. But hey, fine with keeping this :)

**Comment by florian-kuebler** [comment]
> > Wow, thanks so much for this. This was on my todo list since our last fixit-week.
> > On minor thing:
> > As you noted, we should really push forward removing the forward references here.
> > Can you create a bug for that and link this in this PR?
> 
> Hmm. I don't think I have touched a single forwarding reference. Can you explain what you're talking about? Adn why should we remove forwarding references? I'm confused. 🙈

Sorry, I wanted to say "forward declaration". You added those for OrbitApp, which is necessary for now. But should have a bug tracking to remember us to remove them.

File: OrbitQt/orbitcodeeditor.cpp (hunk 2 above)
Comment (florian-kuebler) [reply OrbitQt/orbitcodeeditor.cpp:64]: Fine. Thanks.

File: OrbitGl/TracepointTrack.h (hunk 3 above)
Comment (beckerhe) [reply OrbitGl/TracepointTrack.h:10]: Ah sorry. I should have explained that more clearly. Some tracks are inherited from, notabley TimerTrack and EventTrack. But all the tracks without subclasses had the final tag.

**Comment by beckerhe** [comment]
> > > Wow, thanks so much for this. This was on my todo list since our last fixit-week.
> > > On minor thing:
> > > As you noted, we should really push forward removing the forward references here.
> > > Can you create a bug for that and link this in this PR?
> > 
> > 
> > Hmm. I don't think I have touched a single forwarding reference. Can you explain what you're talking about? Adn why should we remove forwarding references? I'm confused. see_no_evil
> 
> Sorry, I wanted to say "forward declaration". You added those for OrbitApp, which is necessary for now. But should have a bug tracking to remember us to remove them.

Ah sorry. Should have gotten that right away. So regarding adding a bug: I thought about this and think this is not feasible or at least not easily done, because like I said in the PR description: It's not about removing the forward-decl, it's about removing the cyclic dependency. And this needs to be tackled differently - for every class. So I think a single bug doesn't cut it. We would need a bug for every class, so some 30-ish additional bugs. I'm not sure if it is worth it creating those.

If we had a mechanical solution for all classes a single bug would work.

File: OrbitGl/TracepointTrack.h (hunk 3 above)
Comment (florian-kuebler) [reply OrbitGl/TracepointTrack.h:10]: Hmm I don't see the final on any Subclass of TimerTrack.

**Comment by florian-kuebler** [comment]
> > > > Wow, thanks so much for this. This was on my todo list since our last fixit-week.
> > > > On minor thing:
> > > > As you noted, we should really push forward removing the forward references here.
> > > > Can you create a bug for that and link this in this PR?
> > > 
> > > 
> > > Hmm. I don't think I have touched a single forwarding reference. Can you explain what you're talking about? Adn why should we remove forwarding references? I'm confused. see_no_evil
> > 
> > 
> > Sorry, I wanted to say "forward declaration". You added those for OrbitApp, which is necessary for now. But should have a bug tracking to remember us to remove them.
> 
> Ah sorry. Should have gotten that right away. So regarding adding a bug: I thought about this and think this is not feasible or at least not easily done, because like I said in the PR description: It's not about removing the forward-decl, it's about removing the cyclic dependency. And this needs to be tackled differently - for every class. So I think a single bug doesn't cut it. We would need a bug for every class, so some 30-ish additional bugs. I'm not sure if it is worth it creating those.
> 
> If we had a mechanical solution for all classes a single bug would work.
> 
> But I'm open for discussion how we could represent that work in a bug.

What issue do you see in having one bug to bundle all? Even each class needs to be tackled differently. The bug is basically, that we don't forget about this. I agree that 30-ish additional bugs would be a bit much.

File: OrbitGl/TracepointTrack.h (hunk 3 above)
Comment (beckerhe) [reply OrbitGl/TracepointTrack.h:10]: Hmmm. Maybe I'm completely mistaken. Let me just remove it.

**Review by dimitry- — APPROVED** [review]
lgtm!

**Comment by beckerhe** [comment]
> > > > > Wow, thanks so much for this. This was on my todo list since our last fixit-week.
> > > > > On minor thing:
> > > > > As you noted, we should really push forward removing the forward references here.
> > > > > Can you create a bug for that and link this in this PR?
> > > > 
> > > > 
> > > > Hmm. I don't think I have touched a single forwarding reference. Can you explain what you're talking about? Adn why should we remove forwarding references? I'm confused. see_no_evil
> > > 
> > > 
> > > Sorry, I wanted to say "forward declaration". You added those for OrbitApp, which is necessary for now. But should have a bug tracking to remember us to remove them.
> > 
> > 
> > Ah sorry. Should have gotten that right away. So regarding adding a bug: I thought about this and think this is not feasible or at least not easily done, because like I said in the PR description: It's not about removing the forward-decl, it's about removing the cyclic dependency. And this needs to be tackled differently - for every class. So I think a single bug doesn't cut it. We would need a bug for every class, so some 30-ish additional bugs. I'm not sure if it is worth it creating those.
> > If we had a mechanical solution for all classes a single bug would work.
> > But I'm open for discussion how we could represent that work in a bug.
> 
> What issue do you see in having one bug to bundle all? Even each class needs to be tackled differently. The bug is basically, that we don't forget about this. I agree that 30-ish additional bugs would be a bit much.

I'm a bit worried that this will be just one bug which will be around forever. Nobody will ever assign this bug to theirselves because the scope is so huge. Do we have similar bugs? If we are already doing this I will be more comfortable to add another bug in the same manner.

**Comment by florian-kuebler** [comment]
> > > > > > Wow, thanks so much for this. This was on my todo list since our last fixit-week.
> > > > > > On minor thing:
> > > > > > As you noted, we should really push forward removing the forward references here.
> > > > > > Can you create a bug for that and link this in this PR?
> > > > > 
> > > > > 
> > > > > Hmm. I don't think I have touched a single forwarding reference. Can you explain what you're talking about? Adn why should we remove forwarding references? I'm confused. see_no_evil
> > > > 
> > > > 
> > > > Sorry, I wanted to say "forward declaration". You added those for OrbitApp, which is necessary for now. But should have a bug tracking to remember us to remove them.
> > > 
> > > 
> > > Ah sorry. Should have gotten that right away. So regarding adding a bug: I thought about this and think this is not feasible or at least not easily done, because like I said in the PR description: It's not about removing the forward-decl, it's about removing the cyclic dependency. And this needs to be tackled differently - for every class. So I think a single bug doesn't cut it. We would need a bug for every class, so some 30-ish additional bugs. I'm not sure if it is worth it creating those.
> > > If we had a mechanical solution for all classes a single bug would work.
> > > But I'm open for discussion how we could represent that work in a bug.
> > 
> > 
> > What issue do you see in having one bug to bundle all? Even each class needs to be tackled differently. The bug is basically, that we don't forget about this. I agree that 30-ish additional bugs would be a bit much.
> 
> I'm a bit worried that this will be just one bug which will be around forever. Nobody will ever assign this bug to theirselves because the scope is so huge. Do we have similar bugs? If we are already doing this I will be more comfortable to add another bug in the same manner.

IIRC, we had such a bug to get rid of the capture globals. I see you fear, that this bug is around forever, but I guess we can add in the description of the bug, that this needs to be tackled in different ways. Then whenever, someone wants to tackle one of the classes, that person creates a child bug for that specific class. WDYT?

**Comment by beckerhe** [comment]
> > > > > > > Wow, thanks so much for this. This was on my todo list since our last fixit-week.
> > > > > > > On minor thing:
> > > > > > > As you noted, we should really push forward removing the forward references here.
> > > > > > > Can you create a bug for that and link this in this PR?
> > > > > > 
> > > > > > 
> > > > > > Hmm. I don't think I have touched a single forwarding reference. Can you explain what you're talking about? Adn why should we remove forwarding references? I'm confused. see_no_evil
> > > > > 
> > > > > 
> > > > > Sorry, I wanted to say "forward declaration". You added those for OrbitApp, which is necessary for now. But should have a bug tracking to remember us to remove them.
> > > > 
> > > > 
> > > > Ah sorry. Should have gotten that right away. So regarding adding a bug: I thought about this and think this is not feasible or at least not easily done, because like I said in the PR description: It's not about removing the forward-decl, it's about removing the cyclic dependency. And this needs to be tackled differently - for every class. So I think a single bug doesn't cut it. We would need a bug for every class, so some 30-ish additional bugs. I'm not sure if it is worth it creating those.
> > > > If we had a mechanical solution for all classes a single bug would work.
> > > > But I'm open for discussion how we could represent that work in a bug.
> > > 
> > > 
> > > What issue do you see in having one bug to bundle all? Even each class needs to be tackled differently. The bug is basically, that we don't forget about this. I agree that 30-ish additional bugs would be a bit much.
> > 
> > 
> > I'm a bit worried that this will be just one bug which will be around forever. Nobody will ever assign this bug to theirselves because the scope is so huge. Do we have similar bugs? If we are already doing this I will be more comfortable to add another bug in the same manner.
> 
> IIRC, we had such a bug to get rid of the capture globals. I see you fear, that this bug is around forever, but I guess we can add in the description of the bug, that this needs to be tackled in different ways. Then whenever, someone wants to tackle one of the classes, that person creates a child bug for that specific class. WDYT?

Alright. I guees we reached a conclusion. This sounds good to me. :slightly_smiling_face: I will create the bug right away.

**Comment by florian-kuebler** [comment]
> > > > > > > > Wow, thanks so much for this. This was on my todo list since our last fixit-week.
> > > > > > > > On minor thing:
> > > > > > > > As you noted, we should really push forward removing the forward references here.
> > > > > > > > Can you create a bug for that and link this in this PR?
> > > > > > > 
> > > > > > > 
> > > > > > > Hmm. I don't think I have touched a single forwarding reference. Can you explain what you're talking about? Adn why should we remove forwarding references? I'm confused. see_no_evil
> > > > > > 
> > > > > > 
> > > > > > Sorry, I wanted to say "forward declaration". You added those for OrbitApp, which is necessary for now. But should have a bug tracking to remember us to remove them.
> > > > > 
> > > > > 
> > > > > Ah sorry. Should have gotten that right away. So regarding adding a bug: I thought about this and think this is not feasible or at least not easily done, because like I said in the PR description: It's not about removing the forward-decl, it's about removing the cyclic dependency. And this needs to be tackled differently - for every class. So I think a single bug doesn't cut it. We would need a bug for every class, so some 30-ish additional bugs. I'm not sure if it is worth it creating those.
> > > > > If we had a mechanical solution for all classes a single bug would work.
> > > > > But I'm open for discussion how we could represent that work in a bug.
> > > > 
> > > > 
> > > > What issue do you see in having one bug to bundle all? Even each class needs to be tackled differently. The bug is basically, that we don't forget about this. I agree that 30-ish additional bugs would be a bit much.
> > > 
> > > 
> > > I'm a bit worried that this will be just one bug which will be around forever. Nobody will ever assign this bug to theirselves because the scope is so huge. Do we have similar bugs? If we are already doing this I will be more comfortable to add another bug in the same manner.
> > 
> > 
> > IIRC, we had such a bug to get rid of the capture globals. I see you fear, that this bug is around forever, but I guess we can add in the description of the bug, that this needs to be tackled in different ways. Then whenever, someone wants to tackle one of the classes, that person creates a child bug for that specific class. WDYT?
> 
> Alright. I guees we reached a conclusion. This sounds good to me. 🙂 I will create the bug right away.

Thanks!

**Comment by beckerhe** [comment]
> > > > > > > > > Wow, thanks so much for this. This was on my todo list since our last fixit-week.
> > > > > > > > > On minor thing:
> > > > > > > > > As you noted, we should really push forward removing the forward references here.
> > > > > > > > > Can you create a bug for that and link this in this PR?
> > > > > > > > 
> > > > > > > > 
> > > > > > > > Hmm. I don't think I have touched a single forwarding reference. Can you explain what you're talking about? Adn why should we remove forwarding references? I'm confused. see_no_evil
> > > > > > > 
> > > > > > > 
> > > > > > > Sorry, I wanted to say "forward declaration". You added those for OrbitApp, which is necessary for now. But should have a bug tracking to remember us to remove them.
> > > > > > 
> > > > > > 
> > > > > > Ah sorry. Should have gotten that right away. So regarding adding a bug: I thought about this and think this is not feasible or at least not easily done, because like I said in the PR description: It's not about removing the forward-decl, it's about removing the cyclic dependency. And this needs to be tackled differently - for every class. So I think a single bug doesn't cut it. We would need a bug for every class, so some 30-ish additional bugs. I'm not sure if it is worth it creating those.
> > > > > > If we had a mechanical solution for all classes a single bug would work.
> > > > > > But I'm open for discussion how we could represent that work in a bug.
> > > > > 
> > > > > 
> > > > > What issue do you see in having one bug to bundle all? Even each class needs to be tackled differently. The bug is basically, that we don't forget about this. I agree that 30-ish additional bugs would be a bit much.
> > > > 
> > > > 
> > > > I'm a bit worried that this will be just one bug which will be around forever. Nobody will ever assign this bug to theirselves because the scope is so huge. Do we have similar bugs? If we are already doing this I will be more comfortable to add another bug in the same manner.
> > > 
> > > 
> > > IIRC, we had such a bug to get rid of the capture globals. I see you fear, that this bug is around forever, but I guess we can add in the description of the bug, that this needs to be tackled in different ways. Then whenever, someone wants to tackle one of the classes, that person creates a child bug for that specific class. WDYT?
> > 
> > 
> > Alright. I guees we reached a conclusion. This sounds good to me. slightly_smiling_face I will create the bug right away.
> 
> Thanks!

Done. http://b/175759190 Also added to the PR description.

**Comment by beckerhe** [comment]
Heads up, I will remove the fixup commits now, force-push and merge after CI has finished.

If you haven't reviewed this yet and still want to do it, please speak up now.

Florian and Dimitry, thanks a lot for the quick review.

File: OrbitGl/FunctionsDataView.h [hunk 4]
```diff
@@ -8,14 +8,17 @@
 #include "DataView.h"
 #include "capture_data.pb.h"
 
+class OrbitApp;
+
 class FunctionsDataView : public DataView {
  public:
-  FunctionsDataView();
+  FunctionsDataView(OrbitApp* app);
```
Comment (dpallotti) [inline OrbitGl/FunctionsDataView.h:15]: Maybe this should be `explicit` now?

File: OrbitGl/LiveFunctionsController.h [hunk 5]
```diff
@@ -13,9 +13,11 @@
 #include "absl/container/flat_hash_map.h"
 #include "capture_data.pb.h"
 
+class OrbitApp;
+
 class LiveFunctionsController {
  public:
-  LiveFunctionsController() : live_functions_data_view_(this) {}
+  LiveFunctionsController(OrbitApp* app) : live_functions_data_view_(this, app), app_{app} {}
```
Comment (dpallotti) [inline OrbitGl/LiveFunctionsController.h:20]: `explicit`?

File: OrbitGl/PresetsDataView.h [hunk 6]
```diff
@@ -10,9 +10,11 @@
 #include "DataView.h"
 #include "preset.pb.h"
 
+class OrbitApp;
+
 class PresetsDataView : public DataView {
  public:
-  PresetsDataView();
+  PresetsDataView(OrbitApp* app);
```
Comment (dpallotti) [inline OrbitGl/PresetsDataView.h:17]: `explicit`?

File: OrbitGl/TracepointsDataView.h [hunk 7]
```diff
@@ -14,9 +14,11 @@
 
 using orbit_grpc_protos::TracepointInfo;
 
+class OrbitApp;
+
 class TracepointsDataView : public DataView {
  public:
-  TracepointsDataView();
+  TracepointsDataView(OrbitApp* app);
```
Comment (dpallotti) [inline OrbitGl/TracepointsDataView.h:21]: `explicit`?

**Review by dpallotti — APPROVED** [review]
Only skimmed through it and added a couple of comments about `explicit`.

Mostly I wanted to say thank you for the change!!

File: OrbitGl/FunctionsDataView.h (hunk 4 above)
Comment (beckerhe) [reply OrbitGl/FunctionsDataView.h:15]: Yesss. Thanks a lot. I missed that in quite a few locations - it seems! Fixing that now.

Comment (beckerhe) [reply OrbitGl/FunctionsDataView.h]: Done. Thanks again!

File: OrbitGl/LiveFunctionsController.h (hunk 5 above)
Comment (beckerhe) [reply OrbitGl/LiveFunctionsController.h]: Done. Thanks!

File: OrbitGl/PresetsDataView.h (hunk 6 above)
Comment (beckerhe) [reply OrbitGl/PresetsDataView.h]: Done. Thanks!

File: OrbitGl/TracepointsDataView.h (hunk 7 above)
Comment (beckerhe) [reply OrbitGl/TracepointsDataView.h]: Done. Thanks!

**Comment by beckerhe** [comment]
> Only skimmed through it and added a couple of comments about `explicit`.
> 
> Mostly I wanted to say thank you for the change!!

I even found another missing explicit! Thanks a lot for pointing this out.
